### ISSUE TITLE
[SPIR-V] Invalid generated SPIR-V when nesting HLSL 2021 `and` intrinsic.

### ISSUE THREAD

**Issue by Mr-Wiseguy** [microsoft/DirectXShaderCompiler#6995]
**Description**
In an attempt to minimize branching in our HLSL 2021 shader code, we have been converting boolean operators to use the `and` intrinsic from HLSL 2021 that doesn't short circuit. However, when nesting two calls to `and`, we get the following error:
`fatal error: generated SPIR-V is invalid: Expected both operands to be of Result Type: LogicalAnd`

**Steps to Reproduce**
https://shader-playground.timjones.io/a2afd625eb79aac130756befae206296
https://godbolt.org/z/vfn3Kxc7W

**Environment**
- Tested latest July 2024 release and latest trunk on godbolt (5704c474)

**Comment by s-perron** [comment]
This is incorrect even with just a single `and` (https://godbolt.org/z/Te48c9zzz). The AST for the call to `and` is:


```
  |         |-ImplicitCastExpr 0x56546c5b52c8 <col:12> 'bool (*)(int, int)' <FunctionToPointerDecay>
  |         | `-DeclRefExpr 0x56546c5b5260 <col:12> 'bool (int, int)' lvalue Function 0x56546c5b5048 'and' 'bool (int, int)'
```

It treats the parameters as integers. However, the spir-v instruction takes bools. Should be an easy fix.
